Lifetime auto extend #54 sometimes causes test to fail

#54 sometimes causes test to fail. Reverting for now.
https://github.com/joanvila/aioredlock/pull/63/checks?check_run_id=466154021
```
=================================== FAILURES ===================================
_______________________ TestAioredlock.test_cancel_lock_ _______________________

self = <ut.test_algorithm.TestAioredlock object at 0x7fe87b5bba30>
lock_manager_redis_patched = (Aioredlock(redis_connections=[{'host': 'localhost', 'port': 6379}], retry_count=3, retry_delay_min=0.1, retry_delay_max=0.3, internal_lock_timeout=1.0), <class 'asynctest.mock.CoroutineMock'>)

    @pytest.mark.asyncio
    async def test_cancel_lock_(self, lock_manager_redis_patched):
        lock_manager, redis = lock_manager_redis_patched
    
        async def mock_set_lock(*args, **kwargs):
            await real_sleep(1)
    
        redis.set_lock = CoroutineMock(side_effect=mock_set_lock)
    
        with pytest.raises(asyncio.TimeoutError):
            await asyncio.wait_for(lock_manager.lock('resource', 1.0), 0.1)
    
        # The exception handling of the cancelled lock is run in bacround and
        # can not be awaited, so we have to sleep untill the unset_lock has done.
        await real_sleep(0.1)
    
        redis.set_lock.assert_called_once_with('resource', ANY, 1.0)
>       redis.unset_lock.assert_called_once_with('resource', ANY)
E       AssertionError: Expected 'mock' to be called once. Called 0 times.

tests/ut/test_algorithm.py:189: AssertionError
```

_Originally posted by @akhilman in https://github.com/joanvila/aioredlock/pull/54#issuecomment-590635576_

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -1,7 +1,19 @@
 # Change Log
 
 All notable changes to this project will be documented in this file.
+##0.4.0 - 2020-01-08
 
+### Added
+- Add watchdogs to extend lock's lifetime automatically
+
+### Changed
+- Aioredlock totally becomes a lock manager, it track locks' state.When Aioredlock is destroyed, it releases all locks it have
+- Remove ``lock_timeout`` and ``drift`` from Aioredlock's initialization parameter list and move them to ``Aioredlock.lock``
+- Add ``internal_lock_timeout`` as default lock_timeout for Aioredlock initialization
+- ``Aioredlock.lock(resource)`` is replaced by ``Aioredlock.lock(resource_name, lock_timeout)`` 
+- ``Aioredlock.extend(lock)`` is replaced by ``Aioredlock.extend(lock, lock_timeout)``
+- Remove lock_timeout from Redis' initialization parameter list
+ 
 ## 0.3.0 - 2019-01-15
 
 ### Added

--- a/README.rst
+++ b/README.rst
@@ -39,7 +39,7 @@ Usage
 
   # Try to acquire the lock:
   try:
-      lock = await lock_manager.lock("resource_name")
+      lock = await lock_manager.lock("resource_name", lock_timeout=10)
   except LockError:
       print('Lock not acquired')
       raise
@@ -49,7 +49,7 @@ Usage
   assert await lock_manager.is_locked("resource_name")
 
   # Extend lifetime of the lock:
-  await lock_manager.extend(lock)
+  await lock_manager.extend(lock, lock_timeout=10)
   # Raises LockError if the lock manager can not extend the lock lifetime
   # on more then half of the Redis instances.
 
@@ -84,8 +84,6 @@ How it works
 The Aioredlock constructor accepts the following optional parameters:
 
 - ``redis_connections``: A list of connections (dictionary of host and port and kwargs for ``aioredis.create_redis_pool()``, or tuple ``(host, port)``, or string Redis URI) where the Redis instances are running.  The default value is ``[{'host': 'localhost', 'port': 6379}]``.
-- ``lock_timeout``: An float (in seconds) representing lock validity period. The default value is ``10.0`` seconds.
-- ``drift``: An float for clock drift compensation. The default value is calculated by ``lock_timeout * 0.01 + 0.002`` seconds.
 - ``retry_count``: An integer representing number of maximum allowed retries to acquire the lock. The default value is ``3`` times.
 - ``retry_delay_min`` and ``retry_delay_max``: Float values representing waiting time (in seconds) before the next retry attempt. The default values are ``0.1`` and ``0.3``, respectively.
 
@@ -102,7 +100,6 @@ In order to clear all the connections with Redis, the lock_manager ``destroy`` m
 To-do
 -----
 
-* Expire the lock valid attribute according to the lock validity in a safe way if possible
 
 .. _redlock: https://redis.io/topics/distlock
 .. _Redis: https://redis.io

--- a/aioredlock/algorithm.py
+++ b/aioredlock/algorithm.py
@@ -194,7 +194,7 @@ class Aioredlock:
                 except asyncio.CancelledError:
                     pass
                 except Exception:
-                    self.log.exception(f"Can not  unlock {lock.resource}")
+                    self.log.exception('Can not unlock "%s"', lock.resource)
 
             self._watchdogs.pop(lock.resource)
 
@@ -232,7 +232,7 @@ class Aioredlock:
                 try:
                     await self.unlock(lock)
                 except Exception:
-                    self.log.exception(f"Can not  unlock {resource}")
+                    self.log.exception('Can not unlock "%s"', resource)
 
         self._locks.clear()
         self._watchdogs.clear()

--- a/aioredlock/algorithm.py
+++ b/aioredlock/algorithm.py
@@ -13,38 +13,17 @@ from aioredlock.redis import Redis
 
 @attr.s
 class Aioredlock:
-
     redis_connections = attr.ib(default=[{'host': 'localhost', 'port': 6379}])
-
-    lock_timeout = attr.ib(default=10.0, converter=float)
-    # Proportional drift time to the length of the lock
-    # See https://redis.io/topics/distlock#is-the-algorithm-asynchronous for more info
-    drift = attr.ib(default=attr.Factory(
-        lambda self: self.lock_timeout * 0.01 + 0.002, takes_self=True
-    ), converter=float)
 
     retry_count = attr.ib(default=3, converter=int)
     retry_delay_min = attr.ib(default=0.1, converter=float)
     retry_delay_max = attr.ib(default=0.3, converter=float)
+    internal_lock_timeout = attr.ib(default=10.0, converter=float)
 
     def __attrs_post_init__(self):
-        self.redis = Redis(self.redis_connections, self.lock_timeout)
-
-    @lock_timeout.validator
-    def _validate_lock_timeout(self, attribute, value):
-        """
-        Validate if lock_timeout is greater than 0
-        """
-        if value <= 0:
-            raise ValueError("Lock timeout must be greater than 0 seconds.")
-
-    @drift.validator
-    def _validate_drift(self, attribute, value):
-        """
-        Validate if drift is greater than 0
-        """
-        if value <= 0:
-            raise ValueError("Drift must be greater than 0 seconds.")
+        self.redis = Redis(self.redis_connections)
+        self._watchdogs = {}
+        self._locks = {}
 
     @retry_count.validator
     def _validate_retry_count(self, attribute, value):
@@ -53,6 +32,14 @@ class Aioredlock:
         """
         if value < 1:
             raise ValueError("Retry count must be greater or equal 1.")
+
+    @internal_lock_timeout.validator
+    def _validate_internal_lock_timeout(self, attribute, value):
+        """
+        Validate if internal_lock_timeout is greater than 0
+        """
+        if value <= 0:
+            raise ValueError("Internal lock_timeout must be greater than 0 seconds.")
 
     @retry_delay_min.validator
     @retry_delay_max.validator
@@ -67,19 +54,13 @@ class Aioredlock:
     def log(self):
         return logging.getLogger(__name__)
 
-    async def lock(self, resource):
-        """
-        Tries to acquire de lock.
-        If the lock is correctly acquired, the valid property of
-        the returned lock is True.
-        In case of fault the LockError exception will be raised
+    async def _set_lock(self, resource, lock_identifier, lease_time):
 
-        :param resource: The string identifier of the resource to lock
-        :return: :class:`aioredlock.Lock`
-        :raises: LockError in case of fault
-        """
-        lock_identifier = str(uuid.uuid4())
         error = RuntimeError('Retry count less then one')
+
+        # Proportional drift time to the length of the lock
+        # See https://redis.io/topics/distlock#is-the-algorithm-asynchronous for more info
+        drift = lease_time * 0.01 + 0.002
 
         try:
             # global try/except to catch CancelledError
@@ -91,12 +72,12 @@ class Aioredlock:
                                            self.retry_delay_max)
                     await asyncio.sleep(delay)
                 try:
-                    elapsed_time = await self.redis.set_lock(resource, lock_identifier)
+                    elapsed_time = await self.redis.set_lock(resource, lock_identifier, lease_time)
                 except LockError as exc:
                     error = exc
                     continue
 
-                if self.lock_timeout - elapsed_time - self.drift <= 0:
+                if lease_time - elapsed_time - drift <= 0:
                     error = LockError('Lock timeout')
                     self.log.debug('Timeout in acquiring the lock "%s"',
                                    resource)
@@ -119,24 +100,76 @@ class Aioredlock:
 
             raise
 
-        return Lock(self, resource, lock_identifier, valid=True)
+    async def _auto_extend(self, lock):
+        """
+        Tries to reset the lock's lifetime to lock_timeout every 0.6*lock_timeout automatically
+        In case of fault the LockError exception will be raised
+        :param lock: :class:`aioredlock.Lock`
+        :raises: LockError in case of fault
+        """
 
-    async def extend(self, lock):
+        await asyncio.sleep(0.6 * self.internal_lock_timeout)
+        try:
+            await self.extend(lock)
+        except Exception:
+            self.log.debug('Error in extending the lock "%s"',
+                           lock.resource)
+
+        self._watchdogs[lock.resource] = asyncio.ensure_future(self._auto_extend(lock))
+
+    async def lock(self, resource, lock_timeout=None):
+        """
+        Tries to acquire de lock.
+        If the lock is correctly acquired, the valid property of
+        the returned lock is True.
+        In case of fault the LockError exception will be raised
+
+        :param resource: The string identifier of the resource to lock
+        :param lock_timeout: Lock's lifetime
+        :return: :class:`aioredlock.Lock`
+        :raises: LockError in case of fault
+        """
+        lock_identifier = str(uuid.uuid4())
+
+        if lock_timeout is not None and lock_timeout <= 0:
+            raise ValueError("Lock timeout must be greater than 0 seconds.")
+
+        lease_time = lock_timeout or self.internal_lock_timeout
+
+        await self._set_lock(resource, lock_identifier, lease_time)
+
+        lock = Lock(self, resource, lock_identifier, lock_timeout, valid=True)
+        if lock_timeout is None:
+            self._watchdogs[lock.resource] = asyncio.ensure_future(self._auto_extend(lock))
+        self._locks[resource] = lock
+
+        return lock
+
+    async def extend(self, lock, lock_timeout=None):
         """
         Tries to reset the lock's lifetime to lock_timeout
         In case of fault the LockError exception will be raised
 
         :param lock: :class:`aioredlock.Lock`
+        :param lock_timeout: extend lock's life time to lock_timeout
         :raises: RuntimeError if lock is not valid
         :raises: LockError in case of fault
         """
-
         self.log.debug('Extending lock "%s"', lock.resource)
 
         if not lock.valid:
             raise RuntimeError('Lock is not valid')
+        if lock_timeout is not None and lock_timeout <= 0:
+            raise ValueError("Lock timeout must be greater than 0 seconds.")
 
-        await self.redis.set_lock(lock.resource, lock.id)
+        new_lease_time = lock_timeout or lock.lock_timeout or self.internal_lock_timeout
+
+        try:
+            await self._set_lock(lock.resource, lock.id, new_lease_time)
+        except Exception:
+            with contextlib.suppress(LockError):
+                await self.unlock(lock)
+            raise
 
     async def unlock(self, lock):
         """
@@ -147,13 +180,26 @@ class Aioredlock:
         :param lock: :class:`aioredlock.Lock`
         :raises: LockError in case of fault
         """
-
         self.log.debug('Releasing lock "%s"', lock.resource)
+
+        lock.valid = False
+
+        if lock.resource in self._watchdogs:
+            self._watchdogs[lock.resource].cancel()
+
+            done, _ = await asyncio.wait([self._watchdogs[lock.resource]])
+            for fut in done:
+                try:
+                    await fut
+                except asyncio.CancelledError:
+                    pass
+                except Exception:
+                    self.log.exception(f"Can not  unlock {lock.resource}")
+
+            self._watchdogs.pop(lock.resource)
 
         await self.redis.unset_lock(lock.resource, lock.id)
         # raises LockError if can not unlock
-
-        lock.valid = False
 
     async def is_locked(self, resource_or_lock):
         """
@@ -177,7 +223,18 @@ class Aioredlock:
 
     async def destroy(self):
         """
-        Clear all the redis connections
+        cancel all _watchdogs, unlock all locks and Clear all the redis connections
         """
         self.log.debug('Destroying %s', repr(self))
+
+        for resource, lock in self._locks.items():
+            if lock.valid:
+                try:
+                    await self.unlock(lock)
+                except Exception:
+                    self.log.exception(f"Can not  unlock {resource}")
+
+        self._locks.clear()
+        self._watchdogs.clear()
+
         await self.redis.clear_connections()

--- a/aioredlock/lock.py
+++ b/aioredlock/lock.py
@@ -7,6 +7,7 @@ class Lock:
     lock_manager = attr.ib()
     resource = attr.ib()
     id = attr.ib()
+    lock_timeout = attr.ib(default=10.0)
     valid = attr.ib(default=False)
 
     async def __aenter__(self):

--- a/aioredlock/redis.py
+++ b/aioredlock/redis.py
@@ -212,31 +212,29 @@ class Instance:
 
 class Redis:
 
-    def __init__(self, redis_connections, lock_timeout):
+    def __init__(self, redis_connections):
 
         self.instances = []
         for connection in redis_connections:
             self.instances.append(Instance(connection))
 
-        self.lock_timeout = lock_timeout
-
     @property
     def log(self):
         return logging.getLogger(__name__)
 
-    async def set_lock(self, resource, lock_identifier):
+    async def set_lock(self, resource, lock_identifier, lock_timeout=10.0):
         """
         Tries to set the lock to all the redis instances
 
         :param resource: The resource string name to lock
         :param lock_identifier: The id of the lock. A unique string
+        :param lock_timeout: lock's lifetime
         :return float: The elapsed time that took to lock the instances
             in seconds
         :raises: LockError if the lock has not been set to at least (N/2 + 1)
             instances
         """
         start_time = time.time()
-        lock_timeout = self.lock_timeout
 
         successes = await asyncio.gather(*[
             i.set_lock(resource, lock_identifier, lock_timeout) for

--- a/examples/lock_context.py
+++ b/examples/lock_context.py
@@ -16,12 +16,18 @@ async def lock_context():
         print('The resource is already acquired')
 
     try:
+        # if you dont set your lock's lock_timeout, its lifetime will be automatically extended
         async with await lock_manager.lock("resource") as lock:
             assert lock.valid is True
             assert await lock_manager.is_locked("resource") is True
             # Do your stuff having the lock
+            await asyncio.sleep(lock_manager.internal_lock_timeout * 2)
+            # lock manager will extend the lock automatically
+            assert await lock_manager.is_locked(lock)
+            # or you can extend your lock's lifetime manually
             await lock.extend()
-            # Do more stuff having the lock
+            # Do more stuff having the lock and if you spend much more time than you expected, the lock might be freed
+
         assert lock.valid is False  # lock will be released by context manager
     except LockError:
         print('"resource" key might be not empty. Please call '

--- a/setup.py
+++ b/setup.py
@@ -11,7 +11,7 @@ with open(path.join(here, 'README.rst'), encoding='utf-8') as f:
 setup(
     name='aioredlock',
 
-    version='0.3.0',
+    version='0.4.0',
 
     description='Asyncio implemetation of Redis distributed locks',
     long_description=long_description,

--- a/tests/ut/conftest.py
+++ b/tests/ut/conftest.py
@@ -13,7 +13,7 @@ async def dummy_sleep(seconds):
 
 @pytest.fixture
 def locked_lock():
-    return Lock(None, "resource_name", 1, True)
+    return Lock(None, "resource_name", 1, -1, True)
 
 
 @pytest.fixture
@@ -25,7 +25,7 @@ def lock_manager_redis_patched():
             mock_redis.is_locked = CoroutineMock(return_value=False)
             mock_redis.clear_connections = CoroutineMock()
 
-            lock_manager = Aioredlock(lock_timeout=1.0, drift=0.102)
+            lock_manager = Aioredlock(internal_lock_timeout=1.0)
 
             yield lock_manager, mock_redis
 

--- a/tests/ut/test_algorithm.py
+++ b/tests/ut/test_algorithm.py
@@ -333,7 +333,7 @@ class TestAioredlock:
             lock = await lock_manager.lock("resource")
             await real_sleep(lock_manager.internal_lock_timeout)
 
-            if sys.version_info.major == 3 and sys.version_info.minor <= 5:
+            if sys.version_info.major == 3 and sys.version_info.minor <= 6:
                 tasks = asyncio.Task.all_tasks()
             else:
                 tasks = asyncio.all_tasks()

--- a/tests/ut/test_algorithm.py
+++ b/tests/ut/test_algorithm.py
@@ -1,6 +1,7 @@
 import asyncio
 from unittest.mock import ANY, call
 
+import asynctest
 import pytest
 from asynctest import CoroutineMock, patch
 
@@ -10,10 +11,9 @@ real_sleep = asyncio.sleep
 
 
 @pytest.mark.parametrize('method,exc_message', [
-    ('_validate_lock_timeout', "Lock timeout must be greater than 0 seconds."),
-    ('_validate_drift', "Drift must be greater than 0 seconds."),
     ('_validate_retry_count', "Retry count must be greater or equal 1."),
     ('_validate_retry_delay', "Retry delay must be greater than 0 seconds."),
+    ('_validate_internal_lock_timeout', "Internal lock_timeout must be greater than 0 seconds.")
 ])
 def test_validator(method, exc_message):
     with pytest.raises(ValueError) as exc_info:
@@ -30,10 +30,9 @@ class TestAioredlock:
 
             mock_redis.assert_called_once_with(
                 [{'host': 'localhost', 'port': 6379}],
-                lock_manager.lock_timeout
+
             )
             assert lock_manager.redis
-            assert lock_manager.drift == pytest.approx(0.102)
 
     def test_initialization_with_params(self):
         with patch("aioredlock.algorithm.Redis.__init__") as mock_redis:
@@ -42,17 +41,14 @@ class TestAioredlock:
 
             mock_redis.assert_called_once_with(
                 [{'host': '::1', 'port': 1}],
-                lock_manager.lock_timeout
             )
             assert lock_manager.redis
-            assert lock_manager.drift == pytest.approx(0.102)
 
     @pytest.mark.parametrize('param', [
-        'lock_timeout',
-        'drift',
         'retry_count',
         'retry_delay_min',
         'retry_delay_max',
+        'internal_lock_timeout'
     ])
     @pytest.mark.parametrize('value,exc_type', [
         (-1, ValueError),
@@ -70,15 +66,22 @@ class TestAioredlock:
     async def test_lock(self, lock_manager_redis_patched, locked_lock):
         lock_manager, redis = lock_manager_redis_patched
 
-        lock = await lock_manager.lock('resource')
+        lock = await lock_manager.lock('resource', 1.0)
 
         redis.set_lock.assert_called_once_with(
             'resource',
-            ANY
+            ANY,
+            1.0
         )
         assert lock.resource == 'resource'
         assert lock.id == ANY
         assert lock.valid is True
+
+    @pytest.mark.asyncio
+    async def test_lock_with_invalid_param(self, lock_manager_redis_patched):
+        lock_manager, redis = lock_manager_redis_patched
+        with pytest.raises(ValueError):
+            await lock_manager.lock("resource", -1)
 
     @pytest.mark.asyncio
     async def test_lock_one_retry(self, lock_manager_redis_patched, locked_lock):
@@ -88,11 +91,11 @@ class TestAioredlock:
             0.001
         ])
 
-        lock = await lock_manager.lock('resource')
+        lock = await lock_manager.lock('resource', 1.0)
 
         calls = [
-            call('resource', ANY),
-            call('resource', ANY)
+            call('resource', ANY, 1.0),
+            call('resource', ANY, 1.0)
         ]
         redis.set_lock.assert_has_calls(calls)
         redis.unset_lock.assert_not_called()
@@ -110,14 +113,14 @@ class TestAioredlock:
         ])
 
         with pytest.raises(LockError):
-            await lock_manager.lock('resource')
+            await lock_manager.lock('resource', 1.0)
 
         await real_sleep(0.1)  # wait until cleaning is completed
 
         calls = [
-            call('resource', ANY),
-            call('resource', ANY),
-            call('resource', ANY)
+            call('resource', ANY, 1.0),
+            call('resource', ANY, 1.0),
+            call('resource', ANY, 1.0)
         ]
         redis.set_lock.assert_has_calls(calls)
         redis.unset_lock.assert_called_once_with('resource', ANY)
@@ -130,11 +133,11 @@ class TestAioredlock:
             0.001
         ])
 
-        lock = await lock_manager.lock('resource')
+        lock = await lock_manager.lock('resource', 1.0)
 
         calls = [
-            call('resource', ANY),
-            call('resource', ANY)
+            call('resource', ANY, 1.0),
+            call('resource', ANY, 1.0)
         ]
         redis.set_lock.assert_has_calls(calls)
         redis.unset_lock.assert_not_called()
@@ -152,36 +155,14 @@ class TestAioredlock:
         ])
 
         with pytest.raises(LockError):
-            await lock_manager.lock('resource')
+            await lock_manager.lock('resource', 1.0)
 
         await real_sleep(0.1)  # wait until cleaning is completed
 
         calls = [
-            call('resource', ANY),
-            call('resource', ANY),
-            call('resource', ANY)
-        ]
-        redis.set_lock.assert_has_calls(calls)
-        redis.unset_lock.assert_called_once_with('resource', ANY)
-
-    @pytest.mark.asyncio
-    async def test_lock_expire_retries_because_drift(self, lock_manager_redis_patched, locked_lock):
-        lock_manager, redis = lock_manager_redis_patched
-        redis.set_lock = CoroutineMock(side_effect=[
-            0.898,
-            0.970,
-            0.900
-        ])
-
-        with pytest.raises(LockError):
-            await lock_manager.lock('resource')
-
-        await real_sleep(0.1)  # wait until cleaning is completed
-
-        calls = [
-            call('resource', ANY),
-            call('resource', ANY),
-            call('resource', ANY)
+            call('resource', ANY, 1.0),
+            call('resource', ANY, 1.0),
+            call('resource', ANY, 1.0)
         ]
         redis.set_lock.assert_has_calls(calls)
         redis.unset_lock.assert_called_once_with('resource', ANY)
@@ -192,30 +173,29 @@ class TestAioredlock:
 
         async def mock_set_lock(*args, **kwargs):
             await real_sleep(1)
-            return 1.0
 
         redis.set_lock = CoroutineMock(side_effect=mock_set_lock)
 
         with pytest.raises(asyncio.TimeoutError):
-            await asyncio.wait_for(lock_manager.lock('resource'), 0.1)
+            await asyncio.wait_for(lock_manager.lock('resource', 1.0), 0.1)
 
         # The exception handling of the cancelled lock is run in bacround and
         # can not be awaited, so we have to sleep untill the unset_lock has done.
         await real_sleep(0.1)
 
-        redis.set_lock.assert_called_once_with('resource', ANY)
+        redis.set_lock.assert_called_once_with('resource', ANY, 1.0)
         redis.unset_lock.assert_called_once_with('resource', ANY)
 
     @pytest.mark.asyncio
     async def test_extend_lock(self, lock_manager_redis_patched, locked_lock):
         lock_manager, redis = lock_manager_redis_patched
 
-        lock = await lock_manager.lock('resource')
+        lock = await lock_manager.lock('resource', 1.0)
         await lock_manager.extend(lock)
 
         calls = [
-            call('resource', ANY),
-            call('resource', ANY)
+            call('resource', ANY, 1.0),
+            call('resource', ANY, 1.0)
         ]
         redis.set_lock.assert_has_calls(calls)
 
@@ -225,6 +205,23 @@ class TestAioredlock:
 
         await lock_manager.unlock(lock)
         with pytest.raises(RuntimeError):
+            await lock_manager.extend(lock)
+
+    @pytest.mark.asyncio
+    async def test_extend_with_invalid_param(self, lock_manager_redis_patched):
+        lock_manager, redis = lock_manager_redis_patched
+        lock = await lock_manager.lock("resource", 1.0)
+        with pytest.raises(ValueError):
+            await lock_manager.extend(lock, -1)
+
+    @pytest.mark.asyncio
+    async def test_extend_lock_error(self, lock_manager_redis_patched, locked_lock):
+        lock_manager, redis = lock_manager_redis_patched
+        lock = await lock_manager.lock('resource')
+
+        redis.set_lock = CoroutineMock(side_effect=LockError('Can not lock'))
+
+        with pytest.raises(LockError):
             await lock_manager.extend(lock)
 
     @pytest.mark.asyncio
@@ -266,7 +263,7 @@ class TestAioredlock:
     async def test_context_manager(self, lock_manager_redis_patched):
         lock_manager, redis = lock_manager_redis_patched
 
-        async with await lock_manager.lock('resource') as lock:
+        async with await lock_manager.lock('resource', 1.0) as lock:
             assert lock.resource == 'resource'
             assert lock.id == ANY
             assert lock.valid is True
@@ -275,8 +272,8 @@ class TestAioredlock:
         assert lock.valid is False
 
         calls = [
-            call('resource', ANY),
-            call('resource', ANY)
+            call('resource', ANY, 1.0),
+            call('resource', ANY, 1.0)
         ]
         redis.set_lock.assert_has_calls(calls)
         redis.unset_lock.assert_called_once_with('resource', ANY)
@@ -284,7 +281,62 @@ class TestAioredlock:
     @pytest.mark.asyncio
     async def test_destroy_lock_manager(self, lock_manager_redis_patched):
         lock_manager, redis = lock_manager_redis_patched
+        lock_manager.unlock = CoroutineMock(side_effect=LockError('Can not lock'))
 
+        await lock_manager.lock("resource", 1.0)
         await lock_manager.destroy()
 
         redis.clear_connections.assert_called_once_with()
+
+    @pytest.mark.asyncio
+    async def test_auto_extend(self):
+        with asynctest.patch("aioredlock.algorithm.Redis", CoroutineMock) as mock_redis:
+            mock_redis.set_lock = CoroutineMock(return_value=0.005)
+            mock_redis.unset_lock = CoroutineMock(return_value=0.005)
+            mock_redis.clear_connections = CoroutineMock()
+
+            lock_manager = Aioredlock(internal_lock_timeout=1)
+            lock = await lock_manager.lock("resource")
+
+            await real_sleep(lock_manager.internal_lock_timeout * 3)
+
+            calls = [call('resource', lock.id, lock_manager.internal_lock_timeout)
+                     for _ in range(5)]
+            mock_redis.set_lock.assert_has_calls(calls)
+
+            await lock_manager.destroy()
+            mock_redis.clear_connections.assert_called_once_with()
+
+    @pytest.mark.asyncio
+    async def test_auto_extend_with_extend_failed(self):
+        with asynctest.patch("aioredlock.algorithm.Redis", CoroutineMock) as mock_redis:
+            mock_redis.set_lock = CoroutineMock(return_value=0.005)
+            mock_redis.unset_lock = CoroutineMock(return_value=0.005)
+            mock_redis.clear_connections = CoroutineMock()
+
+            lock_manager = Aioredlock(internal_lock_timeout=1.0)
+            lock = await lock_manager.lock("resource")
+            lock.valid = False
+            await real_sleep(lock_manager.internal_lock_timeout * 3)
+            calls = [call('resource', lock.id, lock_manager.internal_lock_timeout)]
+            mock_redis.set_lock.assert_has_calls(calls)
+
+    @pytest.mark.asyncio
+    async def test_unlock_with_watchdog_failed(self):
+        with asynctest.patch("aioredlock.algorithm.Redis", CoroutineMock) as mock_redis:
+            mock_redis.set_lock = CoroutineMock(return_value=0.005)
+            mock_redis.unset_lock = CoroutineMock(return_value=0.005)
+            mock_redis.clear_connections = CoroutineMock()
+            lock_manager = Aioredlock(internal_lock_timeout=1.0)
+
+            lock = await lock_manager.lock("resource")
+            await real_sleep(lock_manager.internal_lock_timeout)
+
+            tasks = asyncio.all_tasks()
+            for index, task in enumerate(tasks):
+                if "_auto_extend" in str(task):
+                    auto_frame = task.get_stack()[-1]
+                    auto_frame.clear()
+
+            await lock_manager.unlock(lock)
+            assert lock.valid is False

--- a/tests/ut/test_algorithm.py
+++ b/tests/ut/test_algorithm.py
@@ -1,4 +1,5 @@
 import asyncio
+import sys
 from unittest.mock import ANY, call
 
 import asynctest
@@ -332,7 +333,10 @@ class TestAioredlock:
             lock = await lock_manager.lock("resource")
             await real_sleep(lock_manager.internal_lock_timeout)
 
-            tasks = asyncio.all_tasks()
+            if sys.version_info.major == 3 and sys.version_info.minor <= 5:
+                tasks = asyncio.Task.all_tasks()
+            else:
+                tasks = asyncio.all_tasks()
             for index, task in enumerate(tasks):
                 if "_auto_extend" in str(task):
                     auto_frame = task.get_stack()[-1]

--- a/tests/ut/test_lock.py
+++ b/tests/ut/test_lock.py
@@ -7,7 +7,7 @@ class TestLock:
 
     def test_lock(self):
         lock_manager = Aioredlock()
-        lock = Lock(lock_manager, "potato", 1)
+        lock = Lock(lock_manager, "potato", 1, 1.0)
         assert lock.resource == "potato"
         assert lock.id == 1
         assert lock.valid is False

--- a/tests/ut/test_redis.py
+++ b/tests/ut/test_redis.py
@@ -194,7 +194,7 @@ def redis_three_connections():
 @pytest.fixture
 def mock_redis_two_instances(redis_two_connections):
     pool = FakePool()
-    redis = Redis(redis_two_connections, 10)
+    redis = Redis(redis_two_connections)
 
     for instance in redis.instances:
         instance._pool = pool
@@ -205,7 +205,7 @@ def mock_redis_two_instances(redis_two_connections):
 @pytest.fixture
 def mock_redis_three_instances(redis_three_connections):
     pool = FakePool()
-    redis = Redis(redis_three_connections, 10)
+    redis = Redis(redis_three_connections)
 
     for instance in redis.instances:
         instance._pool = pool
@@ -219,7 +219,7 @@ class TestRedis:
         with patch("aioredlock.redis.Instance.__init__") as mock_instance:
             mock_instance.return_value = None
 
-            redis = Redis(redis_two_connections, 0.01)
+            redis = Redis(redis_two_connections)
 
             calls = [
                 call({'host': 'localhost', 'port': 6379}),
@@ -227,7 +227,6 @@ class TestRedis:
             ]
             mock_instance.assert_has_calls(calls)
             assert len(redis.instances) == 2
-            assert redis.lock_timeout == 0.01
 
     parametrize_methods = pytest.mark.parametrize("method_name, call_args", [
         ('set_lock', {'keys': ['resource'], 'args':['lock_id', 10000]}),

